### PR TITLE
[RELAY][BACKEND] CompileEngine refactor.

### PR DESCRIPTION
**Pull request by tqchen**
This PR attempts to clean up the interpreter and graph runtime compile by making interface consistent, and organize them into backend modules. It also added support of a CompileEngine that caches previously compiled the code and can be used as the backbone of AOT AND JIT. After this PR, I feel most of the compiler scaffolding is done for the graph compiler support.

- [x] Move interpreter, graph runtime codegen to backend
- [x] Introduce CompileEngine 
- [x] refactor Interpreter and graph runtime codegen to use CompileEngine
- [x] change fuse_ops to create function inplace(not add to globalvar)
- [x] Enhanced graph codegen to support tuple input/output case 
    - The original graph format can be viewed as always returning a tuple
- [x] Make relay.build and relay.optimize consistent with nnvm.
- [x] Change executor interface to also take in ctx and target
   - Allows interpreter to optionally run gpu code

### Notes on some observations I made during the refactor
- The graph runtime codegen need to implement a memory allocator pass to enable reuse of memory.
   - Will leave as a followup 
- The current interpreter don't memoize(and it is important not to because single Expr can evaluate to the different value given different input)
   - It will cause explosive computation in pure data flow form.
   - so it is important to run interpreter in after let binding transformation.
   - We will likely need a liveness analysis to eagerly free memory in the interpreter if we hope to use it to run large networks.
- It might be a good time to think about what would heterogeneous device support looks like as per (cc @zhiics)

### CODE REVIEW

**Comment by tqchen** [comment]
cc @jroesch @eqy @MarisaKirisame  please review

**Comment by jroesch** [comment]
I will review first thing in the morning.

**Comment by jroesch** [comment]
Overall looks good 👍, I think we should still lift all local functions into global definitions before optimizing, it is a good invariant and allows us to do global transformations (in follow-up passes) without having to rewrite the functions which depend on them. If we are worried about performance we can re-perform inning at the end. 

@MarisaKirisame could you take a look at this? it effect us.

**Comment by tqchen** [comment]
re global function lifting. I did not do them in the fuseops to make sure that the current API do not depend heavily on module, once we introduce the entry_func field into module I think we will be more comfortable using modules, that can come as a followup

File: python/tvm/relay/build_module.py [hunk 1]
```diff
@@ -2,45 +2,257 @@
...
+        func = ir_pass.fuse_ops(func)
+        # Graph code generation
+        func = ir_pass.infer_type(func)
+        graph_gen = _graph_gen.GraphRuntimeCodegen(mod=None, target=target)
+        graph_json, lowered_funcs = graph_gen.codegen(func)
+        mod = _tvm_build_module(lowered_funcs, target, target_host=target_host)
     return graph_json, mod, params
+
+
+class GraphExecutor(_interpreter.Executor):
```
[the remark this answers is not in the document]
Comment (tqchen) [reply python/tvm/relay/build_module.py]: Mainly because the logic of graph build is here, and we can reuse the optimize function

**Comment by MarisaKirisame** [comment]
@jroesch how do we do closure? If all function are global I dont think we even have the concept of closure, as we have no partial application.

File: python/tvm/relay/backend/graph_runtime_codegen.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+            if isinstance(arg.checked_type, TupleType):
+                tuple_arg_count += 1
+            inputs.append(self.visit(arg))
+        # We need to specially handle tuple inputs and
+        # tuple output cases.
+        # Tuple input function(e.g. concat)
+        if tuple_arg_count:
+            assert len(call.args) == 1
+            assert isinstance(inputs, tuple)
+            inputs = list(inputs)
```
Comment (MarisaKirisame) [inline python/tvm/relay/backend/graph_runtime_codegen.py]: inputs = list(inputs[0])?

**Comment by MarisaKirisame** [comment]
Overall pretty good, except that line which I think is incorrect.

File: python/tvm/relay/backend/graph_runtime_codegen.py (hunk 2 above)
Comment (tqchen) [reply python/tvm/relay/backend/graph_runtime_codegen.py]: Good catch, fixed by followup PR

**Comment by tqchen** [comment]
@jroesch @MarisaKirisame please take another look, I have rebased and add support for tuple input(concat)

**Pull request merged**